api/src/main/java/org/apache/iceberg/stats/StatsUtil.java

```diff
@@ -0,0 +1,219 @@
+  static final int METADATA_SPACE_FIELD_ID_START = 2_147_000_000;
+  static final int RESERVED_FIELD_IDS_START = Integer.MAX_VALUE - RESERVED_FIELD_IDS;
+  // support stats for only up to this amount of data field IDs
+  static final int MAX_DATA_FIELD_ID = 1_000_000;
+  static final int MAX_DATA_STATS_FIELD_ID = 200_010_000;
+
+  private StatsUtil() {}
+
+  public static int statsFieldIdForField(int fieldId) {
+    return fieldId >= RESERVED_FIELD_IDS_START
+        ? statsFieldIdForReservedField(fieldId)
+        : statsFieldIdForDataField(fieldId);
+  }
+
+  private static int statsFieldIdForDataField(int fieldId) {
+    long statsFieldId = DATA_SPACE_FIELD_ID_START + NUM_STATS_PER_COLUMN * (long) fieldId;
+    if (fieldId < 0 || fieldId > MAX_DATA_FIELD_ID) {
+      return -1;
+    }
+
+    return (int) statsFieldId;
+  }
+
+  private static int statsFieldIdForReservedField(int fieldId) {
+    int offset = RESERVED_FIELD_IDS - (Integer.MAX_VALUE - fieldId);
+
+    long statsFieldId = METADATA_SPACE_FIELD_ID_START + NUM_STATS_PER_COLUMN * (long) offset;
+    if (statsFieldId < 0 || statsFieldId > RESERVED_FIELD_IDS_START) {
+      // ID overflows
+      return -1;
+    }
+
+    return (int) statsFieldId;
+  }
+
+  public static int fieldIdForStatsField(int statsFieldId) {
+    if (statsFieldId < DATA_SPACE_FIELD_ID_START || statsFieldId % NUM_STATS_PER_COLUMN != 0) {
+      return -1;
+    }
+
+    return statsFieldId < METADATA_SPACE_FIELD_ID_START
+        ? fieldIdForStatsFieldFromDataField(statsFieldId)
+        : fieldIdForStatsFieldFromReservedField(statsFieldId);
+  }
+
+  private static int fieldIdForStatsFieldFromDataField(int statsFieldId) {
+    return Math.max(-1, (statsFieldId - DATA_SPACE_FIELD_ID_START) / NUM_STATS_PER_COLUMN);
+  }
+
+  private static int fieldIdForStatsFieldFromReservedField(int statsFieldId) {
+    return Math.max(
+        -1,
+        statsFieldId
+            - RESERVED_FIELD_IDS
+            + (Integer.MAX_VALUE - statsFieldId)
+            + (statsFieldId - METADATA_SPACE_FIELD_ID_START) / NUM_STATS_PER_COLUMN);
+  }
+
+  public static Types.NestedField contentStatsFor(Schema schema) {
+    ContentStatsSchemaVisitor visitor = new ContentStatsSchemaVisitor();
+    Types.NestedField result = TypeUtil.visit(schema, visitor);
+    if (!visitor.skippedFieldIds.isEmpty()) {
+      LOG.warn("Could not create stats schema for field ids: {}", visitor.skippedFieldIds);
+    }
+
+    return result;
+  }
+
+  private static Types.StructType contentStatsFor(Type type, int id) {
+    return Types.StructType.of(
+        optional(
+            id + VALUE_COUNT_OFFSET,
+            "value_count",
+            Types.LongType.get(),
+            "Total value count, including null and NaN"),
+        optional(
+            id + NULL_VALUE_COUNT_OFFSET,
+            "null_value_count",
+            Types.LongType.get(),
+            "Total null value count"),
+        optional(
```

Comment:
Shouldn't we only produce this for types that have `NaN`?  I think we want to be type specific with respect to the schema we produce (especially if we add other stats for more complex types like geo)